File: LeetCode/046.permutations/permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>res;
        sort(nums.begin(), nums.end());
        do{
            vector<int>tmp;
            for (auto val : nums)   tmp.push_back(val);
            res.push_back(tmp);
        }while(next_permutation(nums.begin(), nums.end()));
        return res;
    }
};

File: LeetCode/046.permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>>res = new ArrayList<>();
        permuteHelper(res, nums, 0, nums.length);
        return res;
    }

    public static void permuteHelper(List<List<Integer>>res, int[] nums, int st, int length){
        if (st == length){
            List<Integer>list = new ArrayList<>(length);
            for (int i = 0; i < length; i++)    list.add(nums[i]);
            res.add(list);
        }
        for (int i = st; i < length; i++){
            swap(nums, st, i);
            permuteHelper(res, nums, st + 1, length);
            swap(nums, st, i);
        }
    }

    public static void swap(int[] nums, int i, int j){
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}